contracts/external/nexus/WrappedCoveredPrincipalToken.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.0;
+
+import { ERC20PermitWithSupply, ERC20Permit } from "../../libraries/ERC20PermitWithSupply.sol";
+import { Authorizable } from "../../libraries/Authorizable.sol";
+import { ITranche } from "../../interfaces/ITranche.sol";
+import { SafeERC20, IERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+contract WrappedCoveredPrincipalToken is ERC20PermitWithSupply, Authorizable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    // Address of the base/underlying token which is used to buy the yield bearing token from the wrapped position.
+    // Ex - Dai is used to buy the yvDai yield bearing token
+    address public immutable baseToken;
+
+    // Enumerable address list, It contains the list of allowed tranches that are covered by this contract
+    // Criteria to choose the tranches are -
+    // a). Tranche should have same underlying/base token (i.e ETH, BTC, USDC).
+    // b). Should have the similar risk profiles.
+    EnumerableSet.AddressSet private _allowedTranches;
```

Comment:
The gas refund changes ruined this for me. Curious what your rationale is for using this instead of a 
`mapping(address -> bool)`?

Reply:
It will tell us all the tranches at once, It is good for the UI